patches/tModLoader/Terraria/ModLoader/ModNet.cs

```diff
@@ -227,12 +229,56 @@ internal static bool SyncClientMods(BinaryReader reader, out bool needsReload)
+					}
+					else {
+						reloadRequiredExplanationEntries.Add(new ReloadRequiredExplanation(2, header.name, matching, $"[c/FFFACD:Switch] to v{header.version}\n(Temporarily upgrade from v{clientMod.Version})"));
+					}
+				}
+				else {
+					toEnable.Add(matching.Name);
+					reloadRequiredExplanationEntries.Add(new ReloadRequiredExplanation(3, header.name, matching, $"[c/98FB98:Enable]")); 
+				}
+
 				continue;
 			}
 
-			if (downloadModsFromServers && (header.signed || !onlyDownloadSignedMods))
+			if (downloadModsFromServers && (header.signed || !onlyDownloadSignedMods)) {
 				downloadQueue.Enqueue(header);
+				if (modFiles.FirstOrDefault(mod => mod.Name == header.name && mod.Version == header.version) is LocalMod localMod) {
+					// We have the correct mod and version, but hash is different.
+					// We could differentiate between a workshop mod being different and a local mod, which is likely because user is mod dev or playtester.
+					//if(localMod.location == ModLocation.Workshop)
+					//	reloadRequiredExplanationEntries.Add(new Reason(1, header.name, $"[c/00BFFF:Download] v{header.version} ({string.Concat(header.hash[..4].Select(b => b.ToString("x2")))}) from server\n[c/ff0000:(Mod differs from local copy, it may have been edited!)]", localMod));
+					//else
+					reloadRequiredExplanationEntries.Add(new ReloadRequiredExplanation(1, header.name, localMod, $"[c/00BFFF:Download] v{header.version} ({string.Concat(header.hash[..4].Select(b => b.ToString("x2")))}) from server\n(Mod file differs from local copy)"));
+				}
+				else {
+					LocalMod localModMatchingNameOnly = modFiles.Where(mod => mod.Name == header.name).OrderByDescending(mod => mod.Version).FirstOrDefault(); // Technically might show mod icon from .tmod file that won't be selected to load, but this is fine.
+					if (localModMatchingNameOnly == null) {
+						// We don't have the mod.
+						reloadRequiredExplanationEntries.Add(new ReloadRequiredExplanation(1, header.name, null, $"[c/00BFFF:Download] v{header.version} from server"));
+					}
+					else {
+						if (clientMod != null ) {
+							// We have the mod enabled, but not the correct version.
+							if (clientMod.Version > header.version)
+								reloadRequiredExplanationEntries.Add(new ReloadRequiredExplanation(1, header.name, localModMatchingNameOnly, $"[c/00BFFF:Download] v{header.version} from server\n(Temporarily downgrade from v{clientMod.Version})"));
+							else
+								reloadRequiredExplanationEntries.Add(new ReloadRequiredExplanation(1, header.name, localModMatchingNameOnly, $"[c/00BFFF:Download] v{header.version} from server\n(Upgrade from v{clientMod.Version})"));
+						}
+						else {
+							// We have the mod, but not the correct version.
+							reloadRequiredExplanationEntries.Add(new ReloadRequiredExplanation(1, header.name, localModMatchingNameOnly, $"[c/00BFFF:Download] v{header.version} from server"));
+						}
+					}
+
+				}
+			}
 			else
 				blockedList.Add(header);
```

Comment:
scary, a tiny `else` with no `{}` at the end of a big `if`. Perhaps extracting the code which makes the `ReloadRequiredExplanation` for a mod to download is the way to go.